aiven_db_migrate/migrate/errors.py

```diff
@@ -17,5 +18,16 @@ class PGDataDumpFailedError(Exception):
     pass
 
 
+# this is currently use to support a very specific error, but should be employed
+# any time we launch an exception, to let the user have a better understanding
+# about what has happened.
+@enum.unique
+class PGMigrateFailureReason(enum.Enum):
+    cannot_migrate_to_older_server_version = "cannot_migrate_to_older_server_version"
+    unknown = "unknown"
+
+
 class PGMigrateValidationFailedError(Exception):
-    pass
+    def __init__(self, message, *, reason: PGMigrateFailureReason = PGMigrateFailureReason.unknown):
+        super().__init__(message)
+        self.reason = reason
```

Comment:
I think we should use subclasses of `PGMigrateValidationFailedError` to have increasingly
precise error types, this makes it possible to have `except` block depending on the error
condition and not use the `reason` attribute in a nested `if`.

Having specific error classes will also move the error message into the class
itself and avoid this duplication of message and reason in the code generating
the exception.

We shouldn't have a more specific class only for the server version and not the
10 other failure modes, this should cover each failure.

Reply:
First, my thought on this approach, as I hadn't made it explicit:
Is the final error code AND error message as reported to our (aiven-core) users aiven-core's business, or aiven-db-migrate's business? **I think it's aiven-core's business, hence the authoritative logic should lie there**. I wouldn't move the parsing/error message->final error code logic to aiven-db-migrate; it's not such library's business. We may decide to forward aiven-db-migrate error code if we think it's fit, though.

> We shouldn't have a more specific class only for the server version and not the
10 other failure modes, this should cover each failure.

Sure for the "10 other failure modes", but it's outside the the scope of this ticket. "other reasons and better parsing logic will be added later on.". I plan to add a ticket for this improvement. Even if I used subclassing, I would only add this very specific exception right now; I'm trying not to make too many changes to our postgresql code if they're not essential.

> Having specific error classes will also move the error message into the class
itself and avoid this duplication of message and reason in the code generating
the exception.

Which could be useful, or could be a bane; what if I need to use a better (more precise) error message than I could statically enter in an exception class, and I want to set it at runtime?

>  think we should use subclasses of PGMigrateValidationFailedError to have increasingly
precise error types, this makes it possible to have except block depending on the error
condition and not use the reason attribute in a nested if.

My thoughts:
* I don't really see an use case for this right now or in the immediate future. The migration checks failed, and it's a kind of final error; we need to report it to the user and go on. We don't take different actions anywhere depending on the reason of this failure. So, YAGNI.
* We don't need to use it in a nested if; in fact my current aiven mapper (not pushed because it doesn't pass linting until this PR is merged) looks like this:
```class AivenDBMigrateErrorMapper:
    NON_PSYCOPG2_ERROR_MAPPINGS = {
        PGMigrateFailureReason.cannot_migrate_to_older_server_version: ErrorMapping(
            "Cannot migrate to a target PostgreSQL server version older than source version",
            PGMigrationErrorResultCode.cannot_migrate_to_older_server_version,
        )
    }

    def __init__(self):
        self.log = logging.getLogger(self.__class__.__name__)

    def map_error(self, e: PGMigrateValidationFailedError) -> Tuple[str, List[JsonObject]]:
        if isinstance(e.__cause__, psycopg2.Error):
            self.log.exception("error during migration check")
            result, code = pg_utils.parse_pg_connection_error(e.__cause__)
        else:
            mapping = self.NON_PSYCOPG2_ERROR_MAPPINGS.get(
                e.reason, ErrorMapping(str(e), PGMigrationErrorResultCode.unknown)
            )
            result = mapping.output_message
            code = mapping.output_code

        result_codes = [{"code": code.full}]
        return result, result_codes
```
Of course we could use exception classes instead of enums as keys, no problem here, but no advantage either.

* Maybe (and I say maybe, but that's another ticket) it could make sense to create a specific subclass for psycopg2 errors (DriverError, or something like that) to prevent the nasty ```__cause__``` inspection; or we could just flatten all the errors from aiven-db-migrate into an appropriate PGMigrateValidationFailedError with an appropriate result code; but, again, I don't see a value in an exception hierarchy in this case.

My 2c!

Reply:
> First, my thought on this approach, as I hadn't made it explicit: Is the final error code AND error message as reported to our (aiven-core) users aiven-core's business, or aiven-db-migrate's business? **I think it's aiven-core's business, hence the authoritative logic should lie there**. I wouldn't move the parsing/error message->final error code logic to aiven-db-migrate; it's not such library's business. We may decide to forward aiven-db-migrate error code if we think it's fit, though.

Yes, we shouldn't make the final error code or messages here and reuse it as-is in some other places :+1:

> 
> > We shouldn't have a more specific class only for the server version and not the
> > 10 other failure modes, this should cover each failure.
> 
> Sure for the "10 other failure modes", but it's outside the the scope of this ticket. "other reasons and better parsing logic will be added later on.". I plan to add a ticket for this improvement. Even if I used subclassing, I would only add this very specific exception right now; I'm trying not to make too many changes to our postgresql code if they're not essential.

Ok to move supprt for others errors to later :+1:

But, this is possible _because_ we are already using an exception class and not an error code.
You can refine exception in future versions in a backwards compatible way, you cannot do that with an error code.

> 
> > Having specific error classes will also move the error message into the class
> > itself and avoid this duplication of message and reason in the code generating
> > the exception.
> 
> Which could be useful, or could be a bane; what if I need to use a better (more precise) error message than I could statically enter in an exception class, and I want to set it at runtime?

The error message of an exception does not need to be static.
As a bonus, a typed exception can store structured information that is accessible when catching it :
```python
class PgMigrateMissingLanguagesError(PGMigrateValidationFailedError):
    def __init__(self, missing_languages: Collection[str]):
        languages_list = ", ".join(missing_languages)
        super().__init__(f"Languages not installed in target: {languages_list}")
        self.missing_languages = missing_languages
```
This also has the benefit of moving "unimportant" details of the message formatting outside
of the main logic.

> 
> > think we should use subclasses of PGMigrateValidationFailedError to have increasingly
> > precise error types, this makes it possible to have except block depending on the error
> > condition and not use the reason attribute in a nested if.
> 
> My thoughts:
> 
>     * I don't really see an use case for this right now or in the immediate future. The migration checks failed, and it's a kind of final error; we need to report it to the user and go on. We don't take different actions anywhere depending on the reason of this failure. So, YAGNI.
>  [...]

We are not the only users of the API.
I think exception hierarchy are the most common idiom, in Python, in psycopg2, etc.

We can even see the general trend where python now suggests in pep8 to use the exception types instead of errno,
and psycopg2 has added many detailed exception types to make the life of the API consumers easier.

There is a bit of "python is not go" there :smile:

Reply:
I'm not inherently against exception hierarchies. I think they're the right way to go for most **actionable** errors. As an example, I love Spring JDBC way of mapping Java's vague (and checked) SQLException to very specific, clear, actionable runtime exceptions (ConcurrentModificationError, DuplicatePrimaryKeyViolation, ForeignConstraintViolation, et cetera).

But exceptions should be raised for an exceptional code paths. ```PGMigrate.validate()``` is explicitly designed to let the user know, with a good amount of detail, what would go wrong when performing the "real" action (something like a dry-run). In this case, errors are expected; we wouldn't trigger an exceptional path. And, when you return error codes, **you could return multiple errors**, which can be extremely useful for the user experience, so that the user can fix multiple errors in one pass. "Source database is too large AND it's newer than target database AND we don't have enough permissions on the target database user" :heart: . _That's_ how I like validation to work. [^1]

We're currently constrained by the API we're using now. If I were designing aiven-db-migrate API right now, on the contrary, I would make validate() return something like ```success: bool, errors:List[SomeKindOfErrorClassHoldingACodeAndAMessage]``` (or maybe just errors, but I usually like to be extra-clear), and I would make migrate(), on the contrary, raise exceptions (possibly with a hierarchy, if deemed useful).

> But, this is possible because we are already using an exception class and not an error code.
You can refine exception in future versions in a backwards compatible way, you cannot do that with an error code.

If we were using error codes, for an unhandled error we'd be in a situation where ```code=generic``` and we'd be parsing the message in aiven-core to produce the final message and error, and we would add a specific code in aiven-db-migrate to address this ticket. Nothing would prevent us from doing a step-by-step improvement. But I'd argue that if the design was error-code driven, we'd already have proper error codes for all the cases that we commonly encounter, and generic codes would be employed only for "I don't know what happened" cases.

> We are not the only users of the API.

But we're not trying to redesign the library right now. And the lib has 9 stars, so I'd argue that we're probably the only real users for our api.

[^1]: I know that many validation libraries throw exceptions on failed validation. But there're as many that return failures. Both designs exist and have their merits.

Reply:
> But exceptions should be raised for an exceptional code paths.

It's less true in Python, see `StopIteration` :smile:.

> `PGMigrate.validate()` is explicitly designed to let the user know, with a good amount of detail, what would go wrong when performing the "real" action (something like a dry-run). In this case, errors are expected; we wouldn't trigger an exceptional path. And, when you return error codes, **you could return multiple errors**[1](#user-content-fn-1-428262304625bb51fcf2d1b4db48d8c8), which can be extremely useful for the user experience, so that the user can fix multiple errors in one pass. "Source database is too large AND it's newer than target database AND we don't have enough permissions on the target database user" heart . _That's_ how I like validation to work. [2](#user-content-fn-2-428262304625bb51fcf2d1b4db48d8c8)

100% agree on the errors list, that would be much better than a single exception and having to repeat-until-success.

Anyway, if we're staying with error codes and not doing the full change,
I think we can close this PR, we're not gaining much by adding one single error code.